pkg/apis/azuredisk/v1alpha1/types.go

```diff
@@ -87,26 +85,59 @@ type AzVolumeAttachmentList struct {
+// +kubebuilder:printcolumn:name="LastHeatbeatTime",type=integer,JSONPath=`.status.lastHeatbeatTime`,description="Represents the time stamp at which azure persistent volume driver sent a heatbeat."
 type AzDriverNode struct {
 	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	Spec   AzDriverNodeSpec   `json:"spec"`
-	Status AzDriverNodeStatus `json:"status"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec defines the desired state of a AzDriverNode.
+	// Required.
+	Spec   AzDriverNodeSpec   `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// status represents the current state of AzDriverNode.
+	// If this is nil or empty, clients should prefer other nodes
+	// for persistent volume allocations or pod places for pods which use azure persistent volumes.
+	// +optional
+	Status *AzDriverNodeStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
 }
 
-// AzDriverNodeSpec is the spec for a AzDriverNode resource
+// AzDriverNodeSpec is the spec for a AzDriverNode resource.
 type AzDriverNodeSpec struct {
-	UnderlyingNode string `json:"underlyingNode"`
+	// Name of the node which this AzDriverNode represents.
+	// Required.
+	NodeName *string `json:"nodeName" protobuf:"bytes,1,opt,name=nodeName"`
 }
 
-// AzDriverNodeStatus is the status for a AzDriverNode resource
+// AzDriverNodeStatus is the status for a AzDriverNode resource.
 type AzDriverNodeStatus struct {
-	UnderlyingNode int32 `json:"underlyingNode"`
+	// LastHeatbeatTime represents the timestamp when a heatbeat was sent by driver node plugin.
+	// A recent timestamp means that node-plugin is responsive and is communicating to API server.
+	// Clients should not solely reply on LastHeatbeatTime to ascertain node plugin's health state.
+	// +optional
+	LastHeatbeatTime *int64 `json:"lastHeatbeatTime,omitempty" protobuf:"varint,1,opt,name=lastHeatbeatTime"`
+
+	// HealthStatus is the health message posted by node plugin along with the heartbeat
+	// This field is for status\debugging purpose only.
+	// Clients should not use\parse this field to make decisions in code.
+	// +optional
+	HealthStatus *string `json:"healthStatus,omitempty" protobuf:"bytes,2,opt,name=healthStatus"`
+
+	// ReadyForVolumeAllocation tells client wheather the node plug-in is ready for volume allocation.
+	// If status is not present or ReadyForVolumeAllocation, then clients should prefer
+	// other nodes in the clusters for azure persistent volumes\pod placements for pods with azure disks.
+	// +optional
+	ReadyForVolumeAllocation *bool `json:"readyForVolumeAllocation,omitempty" protobuf:"varint,3,opt,name=readyForVolumeAllocation"`
```

Comment:
Could this be inferred from the status?

Reply:
A general recommendation appears to be that we should have a top level status and then some conditions which can be used to deduce additional information.
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

Reply:
I am also thinking that we should have specific flags for faster processing by the extender and then a generic field for future uses.

Reply:
take a look at my change and the comment and let me know if you think we should change this.

Reply:
Looks to be a reasonable approach.